vac/raw/mix.md

```diff
@@ -0,0 +1,658 @@
+  - Must run both the Mix protocol instance and the instance of the libp2p
+    protocol for the message being published.
+
+#### 2.2 Roles Flexibility
+
+A single mix node can play different roles in different paths:
+
+- It can be a sender node for messages it initiates.
+- It can be an intermediary node for messages it is forwarding.
+- It can be an exit node for messages it is disseminating.
+
+#### 2.3 Incentives
+
+To publish an anonymous libp2p message (_e.g.,_ GossipSub, Ping, etc.), nodes
+must run a mix node instance. This requirement serves as an incentive for nodes
+to participate in the mix network, as it allows them to benefit from the
+anonymity features while also contributing to the network's overall anonymity
+and robustness.
+
+#### 2.4 Node Discovery
+
+All mix nodes participate in the discovery process and maintain a list of
+discovered nodes.
+
+- **Bootstrap Nodes**
+
+  i. The network has a set of well-known bootstrap nodes that new mix nodes
+  can connect to when joining the network.\
+  ii. The bootstrap nodes help new mix nodes discover other active mix nodes in
+  the network.
+
+- **Discovery**
+  
+  i. All mix nodes publish their Ethereum Node Records (ENRs) containing:
+
+  ```json
+  {
+    "id": "v4",
+    "multiaddr": "/ip4/192.0.2.1/udp/9000/quic",
+    "ed25519": "0x5a6fcd3e9d6a5e4d5f71e7e5b4cfa9b7b73d9f5f7e9a8b9c5d7f9e8d5a6f7c9e",
+    "mix": "/mix/1.0.0",
+    "supported_protocols": ["ping", "gossipsub"]
+  }
+  ```
+
+  **Field Explanations**
+
+  - `id`: Indicates the ENR format version (_e.g.,_ `"v4"`).
+  - `multiaddr`: The node's multiaddress, including the transport protocol
+    (_e.g.,_ QUIC) and IP address/port.
+  - `ed25519`: The node's Ed25519 public key, used for Sphinx encryption.
+  - `mix`: Indicates the supported Mix protocol version (_e.g.,_ `"/mix/1.0.0"`).
+  - `supported_protocols`: A list of other libp2p protocols supported by the
+    node (_e.g.,_ Ping, GossipSub, etc.).
+  - Additional fields may be included based on the node's requirements.
+  
+  ii. The mix nodes use a peer discovery protocol like WAKU/Discv5:
+  
+  - Connect to a set of bootstrap nodes when joining the network.
+  - Regularly update their list of known peers.
+  - Obtain a random sample of nodes that is representative of the network.
+
+- **Path Selection (Message Senders Only)**
+
+  To send an anonymous message, a mix node performs the following actions:
+  
+  i. Choose a random exit node that supports the required libp2p protocol for
+  the message.\
+  ii. Select remaining L-1 unique mix nodes randomly without replacement from
+  the list of discovered nodes.
+
+- **Forwarding To Next Hop (Intermediary Nodes Only)**
+
+  When a mix node receives an incoming Sphinx packet, it performs the following
+  actions:
+
+  i. Decrypts the packet to obtain the next hop multiaddress\
+  ii. Checks if the next hop is in the list of discovered nodes.\
+  iii. If not, performs discovery for that specific node.\
```

Comment:
Great question! The intermediary node cannot choose any arbitrary node because the Sphinx packets are encrypted to specific nodes, and any arbitrary node would not be able to decrypt them. Notifying the sender through secure backtracking is also not straightforward, as the failure message would need to be encrypted and sent back through the established mix route in reverse. However, each intermediary node only knows the previous node, not the entire path back to the sender.

An easier solution is for the sender to send messages through multiple paths (~3 to 4) when reliability is a concern. In fact, for the Anonymized GossipSub Protocol (AGP) [WiP], it is recommended that the sender chooses 3-4 message paths. I'll add this to the spec.

Reply:
Sounds good, this works, maybe we should have a heuristic weighing the number of paths vs network overhead and reliability of message delivery

Reply:
Agreed! A heuristic that balances the number of paths against network overhead and message delivery reliability is valuable. This can be evaluated through simulations to determine if, for example, using four paths provides sufficient reliability while managing overhead effectively.

However, this aspect could be omitted from the initial implementation. It is something that top-level applications can configure based on their specific requirements for anonymity, reliability, and performance trade-offs. This would allows us to focus on core functionality while providing flexibility for application-specific tuning later on.